enhancements/sig-architecture/138-workload-completion/README.md

```diff
@@ -0,0 +1,316 @@
+containing the status of each completion rule for the manifest. A new field `TTLSecondsAfterFinished` will be
+added to `DeleteOption` which enables garbage collection for the entire manifestwork after all completion rules
+have been satisfied. If a TTL is set with no completion rules, then the manifestwork will never be considered
+finished and therefore never be garbage collected. Likewise if completion rules are set but not TTL, the manifestwork
+will complete but not be automatically deleted.
+
+Once a particular manifest has been marked as completed, it is no longer elligible to be updated or recreated
+by the spoke controller, regardless of the completion status of other manifests in the manifestwork. When all
+manifests with completion rules have completed, the entire manifestwork is considered to be complete (including
+manifests that did NOT specify completion rules), and no more updates will be made.
+
+### Design Details
+
+#### API change
+
+To configure completion of a manifestwork, `completionRules` will be added to `manifestConfigs`
+
+```go
+type ManifestConfigOption struct {
+    ...
+
+    // CompletionRules defines when a manifest should be considered completed. A completed manifest
+    // will no longer recieve updates or be recreated on deletion. If it is not set or empty,
+    // the manifest will not be considered completable.
+    // +optional
+    CompletionRules []CompletionRule `json:"completionRules,omitempty"`
+}
+
+type CompletionRule struct {
+    // Type defines the option of how a manifest should be considered completed.
+    // It can be CEL, or WellKnownCompletions.
+    // If the type is CEL, user should specify the celExpressions field
+    // If the type is WellKnownCompletions, certain common types in k8s.io/api will be considered
+    // completed as defined by hardcoded rules
+    // +kubebuilder:validation:Required
+    // +required
+    Type CompletionType `json:"type"`
+
+    // CelExpressions defines the CEL expressions to be evaluated for completion
+    // +optional
+    CelExpressions []CelCompletionExpressions `json:"celExpressions"`
+}
+
+// +kubebuilder:validation:Enum=WellKnownCompletions;CEL
+type CompletionType string
+
+const (
+    // WellKnownCompletionType represents standard completions for some common types, which
+    // is reflected with a hardcoded rule only for types in k8s.io/api
+    WellKnownCompletionsType CompletionType = "WellKnownCompletions"
+
+    CelCompletionExpressionsType CompletionType = "CEL"
+)
+
+type CelCompletionExpressions struct {
+    // Expression represents the CEL expression to be evaluated on the manifest.
+    // The expression must evaluate to a single value  in the type of integer, bool or string.
+    // If the expression evaluates to a structure, map or slice, the manifest will not be considered complete.
+    // Ref to https://cel.dev/ on how to write CEL
```

Comment:
Not sure if the current CEL selectors work intends to use the kubernetes CEL libraries that extend the standard CEL lib. If so, can link to those docs instead https://kubernetes.io/docs/reference/using-api/cel/

Does look like there are some useful things in there, such as support for k8s Quantity parsing and comparisons.

Reply:
cc @haoqing0110

Reply:
I don't think of a case that needs to compare the resource quantity and mark as complete. Maybe it's just a code level thing to include a specific library in the future when necessary.

Reply:
Probably not resource quantities, but could imagine a CRD using quantity in status and then comparing something like `status.quantity >= spec.quantity`. 

More generally, I do think making CEL in OCM match CEL in standard kubernetes APIs is a good idea for usability and future proofing.

Reply:
I'm not sure if a crd/data structure that does not support protobuf can be used directly as a variable in cel (ref [Use variables in a function](https://codelabs.developers.google.com/codelabs/cel-go#4)).
Also, maybe can point to https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiserver/pkg/cel/library for the libraries we use, semverlib.go is a new library not list in the doc.

Reply:
Manifests are always treated as `unstructured.Unstructured` right? So I think it will work, because object will be `map[string]interface{}` with a cel variable of type `cel.DynType`

Reply:
I was thinking of linking to https://kubernetes.io/docs/reference/using-api/cel/